source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,209 @@
+    // TODO(toddmgreer): Wait for the cache if necessary.
+    insert_->insertBody(
+        data, [](bool) {}, end_stream);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+void CacheFilter::onOkHeaders(Http::HeaderMapPtr&& headers,
+                              std::vector<AdjustedByteRange>&& /*response_ranges*/,
+                              uint64_t content_length, bool has_trailers) {
+  if (!lookup_) {
+    return;
+  }
+  response_has_trailers_ = has_trailers;
+  const bool end_stream = (content_length == 0 && !response_has_trailers_);
+  // TODO(toddmgreer): Calculate age per https://httpwg.org/specs/rfc7234.html#age.calculations
+  headers->addReferenceKey(Http::Headers::get().Age, 0);
+  decoder_callbacks_->encodeHeaders(std::move(headers), end_stream);
+  if (end_stream) {
+    return;
+  }
+  if (content_length > 0) {
+    remaining_body_.emplace_back(0, content_length);
+    getBody();
+  } else {
+    lookup_->getTrailers([self = shared_from_this()](Http::HeaderMapPtr&& trailers) {
+      onTrailersAsync(self, std::move(trailers));
+    });
+  }
+}
+
+void CacheFilter::onUnusableHeaders() {
+  if (lookup_) {
+    decoder_callbacks_->continueDecoding();
+  }
+}
+
+void CacheFilter::onHeadersAsync(const CacheFilterSharedPtr& self, LookupResult&& result) {
+  switch (result.cache_entry_status_) {
+  case CacheEntryStatus::RequiresValidation:
+  case CacheEntryStatus::FoundNotModified:
+  case CacheEntryStatus::UnsatisfiableRange:
+    ASSERT(false); // We don't yet return or support these codes.
+    FALLTHRU;
+  case CacheEntryStatus::Unusable: {
+    self->post([self] { self->onUnusableHeaders(); });
+    return;
+  }
+  case CacheEntryStatus::Ok:
+    self->post([self, headers = result.headers_.release(),
+                response_ranges = std::move(result.response_ranges_),
+                content_length = result.content_length_,
+                has_trailers = result.has_trailers_]() mutable {
+      self->onOkHeaders(absl::WrapUnique(headers), std::move(response_ranges), content_length,
+                        has_trailers);
+    });
+  }
+}
+
+void CacheFilter::getBody() {
+  ASSERT(!remaining_body_.empty(), "No reason to call getBody when there's no body to get.");
+  CacheFilterSharedPtr self = shared_from_this();
+  lookup_->getBody(remaining_body_[0],
+                   [self](Buffer::InstancePtr&& body) { self->onBody(std::move(body)); });
+}
+
+void CacheFilter::onBodyAsync(const CacheFilterSharedPtr& self, Buffer::InstancePtr&& body) {
+  self->post([self, body = body.release()] { self->onBody(absl::WrapUnique(body)); });
+}
+
+// TODO(toddmgreer): Handle downstream backpressure.
+void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+  if (!lookup_) {
+    return;
+  }
+  if (remaining_body_.empty()) {
+    ASSERT(false, "CacheFilter doesn't call getBody unless there's more body to get, so this is a "
+                  "bogus callback.");
+    decoder_callbacks_->resetStream();
+    return;
```

Comment:
Please don't add error handling for cases like this. A plain ASSERT is fine before proceeding. If something can't happen I would just assert it and move on. Same below. If we want to actually defend against bad cache implementations then I would remove the ASSERT(falses) statements throughout and actually handle things, and probably have stats eventually (so add a TODO for that).